### ISSUE TITLE
nan in sparse mv result when I use cpu 

### ISSUE THREAD

**Issue by liuqi45** [pytorch/pytorch#138168]
### 🐛 Describe the bug

When I use cpu to compute A@x where A is a CSR sparse matrix, x is a vector through a function, if the non-zero elements number in A is large enough, then there somstimes exist a nan in the result vector. Howerver such problem did not occur when I use gpu to compute. 
‘’‘python# 
import torch
import random

random.seed(100)
torch.manual_seed(100)
torch.cuda.manual_seed(100)
torch.set_default_device('cpu')

def A_matvec_x(A,x):
    c = torch.mv(A,x)                 #bug is here
    print(c)                              #sometimes there is a nan
    print(c.max())                    #sometimes there is a nan
    

n = 205082                            #control the scale
CSR_IA = [0]
CSR_JA = []

data = torch.zeros(n)
data = data + torch.rand(n) * torch.finfo(data.dtype).eps

x = torch.zeros(n)
b = torch.zeros(n)
for i in range(n):
    temp = []
    for j in range(4):
        q = random.randint(0, n - 1)
        if(q in temp):
            pass
        else:
            CSR_JA.append(random.randint(0, n - 1))
            temp.append(q)

    CSR_IA.append(CSR_IA[i] + 4)

A = torch.sparse_csr_tensor(CSR_IA, CSR_JA, data, size = (n, n))
c = torch.mv(A,x)
print(c.max())                          

A_matvec_x(A,x)’‘’

the result sometimes is '''tensor(0.)
tensor([0., 0., 0.,  ..., 0., nan, 0.])
tensor(nan)'''

somtimes is 
'''
tensor(0.)
tensor([0., 0., 0.,  ..., 0., 0., 0.])
tensor(0.)
'''

### Versions

PyTorch version: 2.1.2+cu121
Is debug build: False
CUDA used to build PyTorch: 12.1
ROCM used to build PyTorch: N/A

OS: Microsoft Windows 11 专业版
GCC version: (x86_64-posix-seh-rev3, Built by MinGW-W64 project) 11.2.0
Clang version: Could not collect
CMake version: Could not collect
Libc version: N/A

Python version: 3.8.18 (default, Sep 11 2023, 13:39:12) [MSC v.1916 64 bit (AMD64)] (64-bit runtime)
Python platform: Windows-10-10.0.22631-SP0
Is CUDA available: True
CUDA runtime version: 11.8.89
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: GPU 0: NVIDIA GeForce RTX 4080 Laptop GPU
Nvidia driver version: 556.13
cuDNN version: C:\Program Files\NVIDIA GPU Computing Toolkit\CUDA\v11.8\bin\cudnn_ops64_9.dll
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture=9
CurrentClockSpeed=2100
DeviceID=CPU0
Family=198
L2CacheSize=14336
L2CacheSpeed=
Manufacturer=GenuineIntel
MaxClockSpeed=2100
Name=13th Gen Intel(R) Core(TM) i7-13700HX
ProcessorType=3
Revision=

Versions of relevant libraries:
[pip3] numpy==1.24.4
[pip3] torch==2.1.2+cu121
[pip3] torchaudio==2.1.2+cu121
[pip3] torchvision==0.16.2+cu121
[conda] numpy                     1.24.4                   pypi_0    pypi
[conda] torch                     2.1.2+cu121              pypi_0    pypi
[conda] torchaudio                2.1.2+cu121              pypi_0    pypi
[conda] torchvision               0.16.2+cu121             pypi_0    pypi
